xdsl/transforms/common_subexpression_elimination.py

```diff
@@ -92,6 +99,23 @@ def pop(self, k: Operation):
         return self._known_ops.pop(OperationInfo(k))
 
 
+def has_other_side_effecting_op_in_between(
+    from_op: Operation, to_op: Operation
+) -> bool:
+    """
+    Returns if there *may* be a 'write' effecting operation between `from_op` and
+    `to_op`.
+    """
+    assert from_op.parent is to_op.parent
+    next_op: Operation | None = from_op
+    while next_op and (next_op is not to_op):
+        effects = get_side_effects_recursively(next_op)
+        if effects is None or (MemoryEffectKind.WRITE in effects):
+            return True
+        next_op = next_op.next_op
+    return False
```

Comment:
nit:
```suggestion
    next_op = from_op
    while next_op is not to_op:
        effects = get_side_effects_recursively(next_op)
        if effects is None or (MemoryEffectKind.WRITE in effects):
            return True
        next_op = next_op.next_op
        assert next_op is not None, "Incorrect order of ops in side-effect search"
    return False
```
since from_op is before to_op, there should never be a None, right?

Reply:
Right, lifted from MLIR

I think point is, this is an internal helper, not mean to be robust against user input but just avoid the extra hot loop test inside.

I don't mind adding it

Reply:
your call, I think it's ok as-is

Reply:
Will add more comments

More generally, I'll also try to just add links to the MLIR implementation and mention inline the intent where I'm not even trying to be Pythonic-first.